File: 268.MissingNumber/missing.cc
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int sum = int(nums.size()) * (int(nums.size()) + 1) / 2;
        for (vector<int>::iterator it = nums.begin(); it != nums.end(); ++it) {
            sum -= *it;
        }
        return sum;
    }

    int missingNumberOpt(vector<int>& nums) {
        int x = int(nums.size());
        for (int i = 0; i < int(nums.size()); i++) {
            x ^= i ^ nums[i];
        }
        return x;
    }
};

File: 268.MissingNumber/missing.py
# -*- coding: utf-8 -*-
#
class Solution(object):
    # 方法1：求和再减
    def missingNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        s = sum(nums)
        n = len(nums)
        return n*(n+1)/2 -s

    # 方法2：a^b^b=a
    def missingNumberOpt(self, nums):
        res = len(nums)
        for i in range(len(nums)):
            res ^= i ^ nums[i]
        return res

if __name__ == '__main__':
    print Solution().missingNumberOpt([3,2,0])
